pilot/pkg/serviceregistry/kube/controller/multicluster.go

```diff
@@ -120,18 +126,49 @@ func NewMulticluster(
 		secretNamespace:       secretNamespace,
 		syncInterval:          opts.GetSyncInterval(),
 		client:                kc,
+		s:                     s,
 	}
 
 	return mc
 }
 
+func (m *Multicluster) Run(stopCh <-chan struct{}) error {
+	// Wait for server shutdown.
+	<-stopCh
+	return m.close()
+}
+
+func (m *Multicluster) close() (err error) {
+	m.m.Lock()
+	m.closing = true
+
+	// Gather all of the member clusters.
+	var clusterIDs []string
+	for clusterID := range m.remoteKubeControllers {
+		clusterIDs = append(clusterIDs, clusterID)
+	}
+	m.m.Unlock()
+
+	// Remove all of the clusters.
+	for _, clusterID := range clusterIDs {
+		err = multierror.Append(err, m.DeleteMemberCluster(clusterID)).ErrorOrNil()
```

Comment:
Cleaning up the k8s controllers will be a bit slow, can we do this in an errgroup?

Reply:
Good point ... fixed.